### PR TITLE
added "skipOnError" ability & minor changes

### PR DESCRIPTION
**Pull request by thenotsoft**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | #29

### CODE REVIEW

File: src/Rule.php [hunk 1]
```diff
@@ -11,10 +11,15 @@
  */
 abstract class Rule
 {
+    const SKIP_ON_ANY_ERROR = 1;
+    const SKIP_ON_ATTRIBUTE_ERROR = 2;
```
Comment (rustamwin) [inline src/Rule.php]:
```suggestion
    public const SKIP_ON_ANY_ERROR = 1;
    public const SKIP_ON_ATTRIBUTE_ERROR = 2;
```

File: src/ResultSet.php [hunk 2]
```diff
@@ -31,6 +35,11 @@ public function addResult(
         }
     }
 
+    public function hasErrors(): bool
```
Comment (rustamwin) [inline src/ResultSet.php]: :+1:

File: src/Rule.php (hunk 1 above)
Comment (thenotsoft) [reply src/Rule.php]: These constants will be removed

File: src/Rule.php [hunk 3]
```diff
@@ -41,22 +47,25 @@ abstract class Rule
      */
     abstract protected function validateValue($value, DataSetInterface $dataSet = null): Result;
 
-    public function setTranslator(TranslatorInterface $translator): self
+    public function withTranslator(TranslatorInterface $translator): self
     {
-        $this->translator = $translator;
-        return $this;
+        $new = clone $this;
```
Comment (kamarton) [inline src/Rule.php]: Do you really need to create a new copy? I don't think there is a use-case. If your goal is cascading, you should:
```php
public function translator(TranslatorInterface $translator): self 
{
  $this->translator = $translator;
  return $this;
}

$validator->translator(...)->translatorDomain(...)->translationLocale(...)->...
```

File: src/Rule.php [hunk 4]
```diff
@@ -73,6 +82,13 @@ public function translateMessage(string $message, array $arguments = []): string
         );
     }
 
+    public function skipOnError(bool $value): self
```
Comment (kamarton) [inline src/Rule.php]: If you have a new instance: `withSkipOnError()`, but I suggest that you do not have a new instance here:
```php
public function skipOnError(bool $value): self
{
        $this->skipOnError = $value;
        return $this;
}
```

File: src/Validator.php [hunk 5]
```diff
@@ -44,11 +44,10 @@ public function __construct(
     public function validate(DataSetInterface $dataSet): ResultSet
     {
         $results = new ResultSet();
-
         foreach ($this->attributeRules as $attribute => $rules) {
             $results->addResult(
                 $attribute,
-                $rules->validate($dataSet->getAttributeValue($attribute))
+                $rules->validate($dataSet->getAttributeValue($attribute), $dataSet)
```
Comment (samdark) [inline src/Validator.php]: This is not related to skip on error but is important bugfix.

File: src/Rule.php (hunk 3 above)
Comment (samdark) [reply src/Rule.php]: @kamarton cloning is cheap. I don't think there's anything bad in making object immutable if it's not meant to have state.

**Comment by samdark** [comment]
👍

**Pull request merged**